internal/jimm/juju/model_poller.go

```
@@ -0,0 +1,145 @@
+	})
+	if err != nil {
+		return errors.E(op, err)
+	}
+
+	// Step 2: Loop over controllers and process their models
+	// This way we only dial each controller once.
+	for controllerUUID, models := range controllerModels {
+		if len(models) == 0 {
+			continue
+		}
+		api, err := j.dialController(ctx, &models[0].Controller)
+		if err != nil {
+			zapctx.Error(ctx, "cannot dial controller", zap.String("controller", controllerUUID), zap.Error(err))
+			continue
+		}
+		// Depending the model's migration mode, we either:
+		// - Check if the model exists (MigrationModeNone)
+		// - Check if the model has completed internal migration (MigrationModeMigrateInternal)
+		// - Do nothing if the model is in any other migration mode (MigrationModeImporting, MigrationModeExporting)
+		for _, m := range models {
+			switch m.MigrationMode {
+			case dbmodel.MigrationModeNone:
+				j.checkModelExists(ctx, api, m)
+			case dbmodel.MigrationModeMigrateInternal:
+				j.checkModelMigratedInternal(ctx, api, m)
+			}
+		}
+	}
+	return nil
+}
+
+// checkModelMigratedInternal checks if the model has been migrated from
+// one controller managed by JIMM to another controller managed by JIMM.
+func (j *JujuManager) checkModelMigratedInternal(ctx context.Context, api API, m *dbmodel.Model) {
+	const op = errors.Op("jimm.checkModelMoved")
+	zapctx.Info(ctx, string(op))
+
+	// Check if the model has completed a migration.
+	// If modelInfo returns without an error, it definitely hasn't moved yet.
+	modelInfo := &jujuparams.ModelInfo{UUID: m.UUID.String}
+	err := api.ModelInfo(ctx, modelInfo)
+	if err == nil {
+		// If the migration end time is set, it means the model has
+		// failed to migrate otherwise we'd expect a redirect error.
+		if modelInfo.Migration.End != nil {
+			m.ProcessFailedMigration()
+			err = j.Database.UpdateModel(ctx, m)
+			if err != nil {
+				zapctx.Error(ctx, "failed to update model after failed migration", zap.String("model", m.UUID.String), zap.Error(err))
```

We can return an error in this function and have a single zapctx.Error in PollModels, there's no reason this couldn't return an error and the caller should define the behaviour of what to do in the event of an error. 

tldr bubble errors up please.